File: kattis/sequences.cs
using System;

public class Program
{
    static long f1(long x, long p) => (x << 1) % p;

    static long f2(long x, int o, long q, long p) => (x + o * q) % p;
    public static void Main()
    {
        const int mod = 1000000007;
        long r = 0;
        int o = 0;
        long q = 500000004;

        string line = Console.ReadLine();
        for (int i = 0; i < line.Length; i++)
        {
            if (line[i] == '?')
            {
                r = f1(r, mod);
                r = f2(r, o, q, mod);

                q = f1(q, mod);

                o += 1;
            }
            else if (line[i] == '0')
            {
                r = f2(r, o, q, mod);
            }
            else
            {
                o += 2;
            }
        }
        Console.WriteLine(r);
    }
}

File: kattis/sequences.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func f1(x int64, p int64) int64{
	return (x << 1) % p
}

func f2(x int64, o int64, q int64, p int64) int64{
	return (x + o * q) % p
}

func main() {
	var mod int64 = 1000000007
	var q int64 = 500000004
	var r int64 = 0
	var o int64 = 0

	line, _ := bufio.NewReader(os.Stdin).ReadString('\n')
	
	for _, c := range line {
		if c == '?' {
			r = f1(r, mod)
			r = f2(r, o, q, mod)

			q = f1(q, mod)

			o += 1
		} else if c == '0' {
			r = f2(r, o, q, mod)
		} else {
			o += 2
		}
	}

	fmt.Printf("%d\n", r)
}
